core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -409,4 +417,50 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    private static class QueueSizeAdjuster
+    {
+        private final Map<String, Long> splitsWeightPerNode;
+        private final long minPendingSplitsWeightPerTask;
+        private final long maxPendingSplitsWeightPerTask;
+        private final long maxSplitsWeightPerNode;
+
+        QueueSizeAdjuster(long minPendingSplitsWeightPerTask, long maxPendingSplitsWeightPerTask, long maxSplitsWeightPerNode)
+        {
+            this.splitsWeightPerNode = new HashMap<>();
+            this.maxPendingSplitsWeightPerTask = maxPendingSplitsWeightPerTask;
+            this.minPendingSplitsWeightPerTask = minPendingSplitsWeightPerTask;
+            this.maxSplitsWeightPerNode = maxSplitsWeightPerNode;
+        }
+
+        public void update(List<RemoteTask> existingTasks, NodeAssignmentStats nodeAssignmentStats)
+        {
+            if (maxPendingSplitsWeightPerTask == minPendingSplitsWeightPerTask) {
+                return;
+            }
+            for (RemoteTask task : existingTasks) {
+                String nodeId = task.getNodeId();
+                long currentSplitsWeightPerNode = getMaxPendingSplitsWeightPerTask(nodeId);
+
+                if (currentSplitsWeightPerNode < maxPendingSplitsWeightPerTask && nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId) == 0) {
+                    currentSplitsWeightPerNode = Math.min(maxPendingSplitsWeightPerTask, currentSplitsWeightPerNode * 2);
```

Comment:
You should only bump it higher when node had `new_max_splits_per_node ` splits queued during previous schedule.
E.g. we want to prevent case when `new_max_splits_per_node  = 20` but connector can only provide 10 splits per batch. In that case, it doesn't make sense to bump `new_max_splits_per_node ` higher (because connector cannot fill the quota anyway)